### PR TITLE
Add dimension mapping parameter

### PR DESCRIPTION
**Pull request by csoulios**
Added the `dimension` parameter to the following field types:
- `keyword`
- `ip`
- Numeric field types (`integer`, `long`, `byte`, `short`)

The `dimension` parameter is of type boolean (default: `false`) and is used to mark that a field is a time series dimension field.

Relates to https://github.com/elastic/elasticsearch/issues/74014

Example of setting `dimension` param to a field can be seen here:

```
PUT my-index-000001
{
  "mappings": {
    "properties": {
      "hostname": {
        "type": "keyword",
        "dimension": true
      },
      "ip_address": {
        "type": "ip",
        "dimension": true
      },
      "host_id": {
        "type": "long",
        "dimension": true
      }
    }
  }
}
```

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java [hunk 1]
```diff
@@ -70,6 +70,21 @@ protected void registerParameters(ParameterChecker checker) throws IOException {
                 },
                 m -> assertThat((m).onScriptError, equalTo("continue")));
         }
+
+        // dimension can be set from false to true, but not vice versa
+        checker.registerUpdateCheck(b -> b.field("dimension", true),
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java:75]: I think the last time we talked we settled on making it fully immutable - no one way transitions. Is that right, @imotov ?

Comment (csoulios) [reply server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java:75]: Yes, that's what we discussed. I am still working on this. The PR is WIP

Comment (imotov) [reply server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java]: Indeed that's the conclusion that we reached. 

My understanding that we are moving away from WIP label towards Draft PRs. I found draft PRs really in your face in terms of distinguishing between WIP and Ready for review PRs 😄  so I typically open all my WIP PRs as Drafts.

Comment (csoulios) [reply server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java]: Converted this to draft PR and removed the `WIP` label. Thanks Igor

File: docs/reference/mapping/types/keyword.asciidoc [hunk 2]
```diff
@@ -56,6 +56,11 @@ The following parameters are accepted by `keyword` fields:
 
 [horizontal]
 
+<<dimension,`dimension`>>::
+
+    This parameter marks the field as a time series dimension field. Accepts
+    `true` or `false` (default).
```
Comment (jrodewig) [inline docs/reference/mapping/types/keyword.asciidoc:62]:
```suggestion
`dimension`::

    Marks the field as a time series dimension. Accepts `true` or `false`
    (default).
```

File: docs/reference/mapping/types/numeric.asciidoc [hunk 3]
```diff
@@ -119,6 +119,12 @@ The following parameters are accepted by numeric types:
     Accepts `true` (default) and `false`. Not applicable for `unsigned_long`.
     Note that this cannot be set if the `script` parameter is used.
 
+<<dimension,`dimension`>>::
+
+    This parameter marks the field as a time series dimension field. Accepts
+    `true` or `false` (default). This parameter can be set only for fields
+    of type `byte`, `short`, `integer` and `long`.
```
Comment (jrodewig) [inline docs/reference/mapping/types/numeric.asciidoc:126]:
```suggestion
`dimension`::

    Marks the field as a time series dimension. Accepts `true` or `false`
    (default). Only `byte`, `short`, `integer`, and `long` numeric fields
    support this parameter.
```

File: docs/reference/mapping/types/ip.asciidoc [hunk 4]
```diff
@@ -45,6 +45,11 @@ The following parameters are accepted by `ip` fields:
 
 [horizontal]
 
+<<dimension,`dimension`>>::
+
+    This parameter marks the field as a time series dimension field. Accepts
+    `true` or `false` (default).
```
Comment (jrodewig) [inline docs/reference/mapping/types/ip.asciidoc:51]: Removing the `<<...>>` syntax should fix the docs build.
```suggestion
`dimension`::

    Marks the field as a time series dimension. Accepts `true` or `false`
    (default).
```

**Review by jrodewig — APPROVED** [review]
LGTM. I left some minor suggestions. Removing the `<<...>>` syntax from the definitions should fix the docs build.

We'll eventually want to provide a definition for "time series dimension," but I can handle that once we've added the related `metric` mapping parameter.

File: server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java [hunk 5]
```diff
@@ -248,6 +249,28 @@ public void testOutOfRangeValues() throws IOException {
...
+        assertThat(fieldType, instanceOf(NumberFieldMapper.NumberFieldType.class));
+        NumberFieldMapper.NumberFieldType ft = (NumberFieldMapper.NumberFieldType) fieldType;
+        assertFalse(ft.isDimension());
+
+        try {
+            createMapperService(fieldMapping(b -> {
+                minimalMapping(b);
+                b.field("dimension", true);
+            }));
+            fail("Mapper parsing exception expected for non-integer numeric with dimension parameter set");
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java:268]: I *think* `expectThrows` would be a little more readable here.

File: server/src/test/java/org/elasticsearch/index/mapper/WholeNumberFieldMapperTests.java [hunk 6]
```diff
@@ -22,4 +24,44 @@ protected void testDecimalCoerce() throws IOException {
...
+            }));
+        checker.registerConflictCheck("dimension",
+            fieldMapping(b -> {
+                minimalMapping(b);
+                b.field("dimension", true);
+            }),
+            fieldMapping(b -> {
+                minimalMapping(b);
+                b.field("dimension", false);
+            }));
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/index/mapper/WholeNumberFieldMapperTests.java:64]: I wonder if we could stick this in a method on `MapperTestCase` or something. That way all of the fields with the `dimension` parameter could add these with a single line. That way its a little more obvious that we expect them to work the same way.

File: server/src/test/java/org/elasticsearch/search/aggregations/AggregatorBaseTests.java [hunk 7]
```diff
@@ -80,7 +80,8 @@ private ValuesSourceConfig getVSConfig(
         AggregationContext context
     ) {
         MappedFieldType ft
-            = new NumberFieldMapper.NumberFieldType(fieldName, numType, indexed, false, true, false, null, Collections.emptyMap(), null);
+            = new NumberFieldMapper.NumberFieldType(fieldName, numType, indexed, false, true, false,
+            null, Collections.emptyMap(), null, false);
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/search/aggregations/AggregatorBaseTests.java:84]: I'd hit this line with the formatter. It'll be super big but that's the formatter we've chosen. One day that formatter's going to format everything and when we have to reflow something like this I figure we may as well just do what it's going to do.

**Review by nik9000 — COMMENTED** [review]
I left a few small comments. I think it's pretty much right though.

File: docs/reference/mapping/types/ip.asciidoc (hunk 4 above)
Comment (csoulios) [reply docs/reference/mapping/types/ip.asciidoc]: Thanks a lot James!

File: server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java (hunk 5 above)
Comment (csoulios) [reply server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java:268]: Right, I merely copied the code in the method above. I fixed both

**Review by imotov — APPROVED** [review]
LGTM. Are we going to backport it to 7.x? If yes, I think we should add more information, an experimental flag or "for internal use only" message to documentation to avoid causing issues for the end users since for a while it will be only used by kibana.

**Comment by csoulios** [comment]
@imotov I was planning to backport this PR to 7.x. Maybe, I can backport it later when the parameter is exposed  through  the field caps api, so it can be set and retrieved.

By "experimental flag" you mean only document the parameter or you think of adding a setting to enable it?

@jrodewig should we mark this feature as "internal use only"? Is there such a practice?

File: test/framework/src/main/java/org/elasticsearch/index/mapper/MapperTestCase.java [hunk 8]
```diff
@@ -528,6 +538,33 @@ protected String randomFetchTestFormat() {
         return null;
     }
 
+    /**
+     * Test that dimension parameter is not updateable
+     */
+    protected void registerDimensionParameter(ParameterChecker checker) throws IOException {
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/index/mapper/MapperTestCase.java:544]: I think this one works well. Maybe rename to `registerDimensionChecks` or something? It's not registering the parameter so much as the checks. I'm not sure.

File: test/framework/src/main/java/org/elasticsearch/index/mapper/MapperTestCase.java [hunk 9]
```diff
@@ -171,6 +171,16 @@ protected static void assertNoDocValuesField(ParseContext.Document doc, String f
         }
     }
 
+    protected void assertDimension(boolean isDimension, Consumer<MappedFieldType> checker) throws IOException {
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/index/mapper/MapperTestCase.java:174]: Yeah, you're right, this one is pretty ugly. If you passed a `Function<MappedFieldType, Boolean>` I think it'd be a little more readable. But still kind of messy to call. We have other code that hides this with an unchecked cast - I'm not sure that's better. But it'd look something like:
```
protected <T> void assertDimension(boolean isDimension, Function<T, Boolean> isDimension) throws IOException {
     MapperService mapperService = ....;
     @SuppressWarnings("unchecked") // Syntactic sugar in tests
     T fieldType = (T) mapperService.fieldType("field");
     assertThat(isDimension.apply(fieldType), equalTo(isDimensions));
}
```

You cal it like:
```
    assertDimension(true, NumberFieldType::isDimension);
```

I think it'd be *shorter* but maybe less clear.

**Review by nik9000 — APPROVED** [review]
I think the `MapperTestCse#registerDimensionParameter` really helps with readability. `MapperTestCase#assertDimension` maybe not so much. I left some ideas but I'd be happy if you wanted to revert that one. Either way.

Other than those two things the code LGTM - I think like @imotov says we probably want to mark the things experimental until we land more code around them.

File: docs/reference/mapping/types/ip.asciidoc [hunk 10]
```diff
@@ -45,6 +45,11 @@ The following parameters are accepted by `ip` fields:
 
 [horizontal]
 
+`dimension`::
+
+    Marks the field as a time series dimension. Accepts `true` or `false`
+    (default).
```
Comment (jrodewig) [inline docs/reference/mapping/types/ip.asciidoc:51]:
```suggestion
`dimension`::

    For internal use only. Marks the field as a time series dimension in {kib}.
    Accepts `true` or `false` (default).
```

File: docs/reference/mapping/types/numeric.asciidoc [hunk 11]
```diff
@@ -119,6 +119,12 @@ The following parameters are accepted by numeric types:
     Accepts `true` (default) and `false`. Not applicable for `unsigned_long`.
     Note that this cannot be set if the `script` parameter is used.
 
+`dimension`::
+
+    Marks the field as a time series dimension. Accepts `true` or `false`
+    (default). Only `byte`, `short`, `integer`, and `long` numeric fields
+    support this parameter.
```
Comment (jrodewig) [inline docs/reference/mapping/types/numeric.asciidoc:126]:
```suggestion
    For internal use only. Marks the field as a time series dimension in {kib}.
    Accepts `true` or `false` (default). Only `byte`, `short`, `integer`, and
    `long` numeric fields support this parameter.
```

File: docs/reference/mapping/types/keyword.asciidoc [hunk 12]
```diff
@@ -56,6 +56,11 @@ The following parameters are accepted by `keyword` fields:
 
 [horizontal]
 
+`dimension`::
+
+    Marks the field as a time series dimension. Accepts `true` or `false`
+    (default).
```
Comment (jrodewig) [inline docs/reference/mapping/types/keyword.asciidoc:62]:
```suggestion
    For internal use only. Marks the field as a time series dimension in {kib}.
    Accepts `true` or `false` (default). If `true`, you cannot specify
    `ignore_above` for the field.
```

**Review by jrodewig — COMMENTED** [review]
> @jrodewig should we mark this feature as "internal use only"? Is there such a practice?

We do this in a few places, such as the [Fleet APIs](https://www.elastic.co/guide/en/elasticsearch/reference/master/fleet-apis.html).  I've added a few suggestions to change the description to:

```
For internal use only. Marks the field as a time series dimension in {kib}.
...
```

Let me know if we decide to release these parameters as experimental. If so, there's some additional syntax we'll want to add.

I also added a suggestion to the `keyword` version noting you cannot use `ignore_above` with `dimension`. Thanks!

**Review by jimczi — APPROVED** [review]
LGTM too.

**Pull request merged**